### ISSUE TITLE
Merges can create Git LFS pointer files with CRLF line endings on Windows

### ISSUE THREAD

**Issue by larsxschneider** [github/git-lfs#1544]
### Setup

```
git-lfs/1.4.1 (GitHub; windows amd64; go 1.7; git 2070e4c)
git version 2.9.2.windows.1 
Win 7 & Win 8.1
```

### Steps to reproduce

```bash
$ git init .
$ touch first.commit
$ git add .
$ git commit -m "first commit"

$ git checkout -b branch
$ git lfs track "*.bin"
$ git add .
$ git commit -m "tracking *.bin files"

$ git checkout master
$ echo "big-data" > file.dat
$ git lfs track "*.dat"
$ git add .
$ git commit -m "tracking *.dat files"

$ git checkout branch
$ git merge master
$ vi .gitattributes       # Combine both lines to fix the merge conflict!
$ git add .
$ git commit -m "merged"

$ git cat-file -p :file.dat | od -c
0000000   v   e   r   s   i   o   n       h   t   t   p   s   :   /   /
0000020   g   i   t   -   l   f   s   .   g   i   t   h   u   b   .   c
0000040   o   m   /   s   p   e   c   /   v   1  \r  \n   o   i   d
0000060   s   h   a   2   5   6   :   0   9   9   e   e   7   8   2   d
0000100   a   2   9   b   0   2   d   c   c   8   9   8   5   1   b   6
0000120   3   9   7   b   f   d   9   8   5   f   a   f   3   7   5   3
0000140   0   f   b   a   9   e   1   1   1   2   7   7   7   1   0   e
0000160   4   0   8   f   2   6   d  \r  \n   s   i   z   e       9  \r
                                    ^^^^^^^^                        ^^^^  
0000200  \n
        ^^^^
0000201
```

### Explanation

If there is a conflict in the `.gitattributes` during a merge then it looks like as if the attributes are not applied (which kind of makes sense as Git would not know what to do). As a result Git LFS pointer files are not filtered and considered `text`. Consequently they end up with native line endings (`CRLF` on Windows!) in the working tree. After resolving the conflict in `.gitattributes` all LFS pointer files are marked as binary (`-text`), again. As a result the `CRLF` line endings are committed on a Windows.

If Git LFS processes these files again (e.g. with a fresh clone) then it will detect CRLF line endings in Git LFS pointer files and fix them. Git will tell the user that these files have been changed. That is confusing as the content did not change at all.

### Solution

Unfortunately, I don't have a good suggestion to fix the problem. Git LFS could print a warning if LFS pointer files with `CRLF` line endings are corrected. This way the user would at least have a clue what is going on.

The only possible solution I could think of is in Git core. We could tell Git to not perform any line ending conversions if there is a merge conflict in the `.gitattributes` file. What do you think @dscho @peff ?

**Comment by larsxschneider** [comment]
A Pull Request with a fix for the line endings in a Git LFS pointer file looks like this:
![screen shot 2016-09-28 at 15 41 53](https://cloud.githubusercontent.com/assets/477434/18915834/a4390e6e-8592-11e6-9f50-5d66f4b85941.png)
Maybe GH could inspect the pointer file, detect that the Git LFS SHA256 is the same and tell the user that no content was changed. Maybe telling the user that Git LFS meta info was changed or something...

**Comment by peff** [comment]
I suspect you could come up with an equally-bad example if you decide not to do line ending conversion due to the conflict in `.gitattributes`. That behavior happens to match your case, but aren't there (totally non-LFS) cases where it would be wrong _not_ to apply a filter?

My intuition is that the right solution would involve salvaging what we could of the `.gitattributes` file when there's a merge conflict, perhaps by using the version from `HEAD` (or more likely, the `ours` stage of the index). That's not _perfect_, but at least it means that an unrelated merge conflict in the attributes can't spread the damage to files whose entries weren't touched.

**Comment by peff** [comment]
Also, I agree that this is a general git-core problem; it's probably worth making a report to the mailing list.

**Comment by ttaylorr** [comment]
> I suspect you could come up with an equally-bad example if you decide not to do line ending conversion due to the conflict in `.gitattributes`.

I agree, this problem definitely goes both ways. I'd be interested to explore putting this potential feature behind a flag, so that callers can specify which behavior they want.

**Comment by larsxschneider** [comment]
I made a report on the Git mailing list here: http://public-inbox.org/git/248A6E81-8D5C-4183-9756-51A0D5193E3E@gmail.com/

**Comment by ttaylorr** [comment]
Thanks for getting the conversation going here, @larsxschneider. I'm happy to make whatever LFS changes are required for this based on the outcome of that mailing list thread. If a consensus is reached before I start wrapping things up for 1.5, I'll make sure that it gets in.

**Comment by larsxschneider** [comment]
Unfortuantly the Git mailing list does not seem to be that interested in this topic. If time allows I will try to propose a patch. As far as I can see no Git LFS adjustment should be necessary if we solve the problem in the way @peff suggested above.

**Comment by dscho** [comment]
@larsxschneider I just figured out that the pre-merge version of the `.gitattributes` is used for the merge, so I think that there was a misunderstanding: the `.gitattributes` are *not* ignored when there are merge conflicts, because they are read *before* any merge happens, and therefore there cannot be any merge conflicts (unless left-over from previous merge attempts).

I send a reply with a little more verbose explanation to your mail.
